### PR TITLE
fix(security): #MVS-114 add resource filter for APIs in TimeSlotController

### PR DESCRIPTION
**Pull request by Gregprod33**
## Describe your changes
- For the route **@Get("/structures/:id/time-slot")** : create Resource Filter that checks the "id" query param that represents the structure ID.
- For The route : **@Put("/time-slots")**: Create Resource filter AdminRightStructure to check structure Id and workflow right : "Viescolaire.view". Also modify the front model TimeSlot.ts to pass structureId.
- For the route : **@Get("/timeslot/audience/:audienceId")** Create resource filter AccessIfMyStructureFromAudience that returns the structureId belonging to the audience id passed in URI and checks whether the user is linked to this structure.

**This filter uses ClasseService within its class. I've tested that this service was not instanciated each time the API route is called by checking the service ID in logs at the first filter instanciation when vertx is launched.
Then, i've checked whether the service id was the same when calling the API. This demonstrates that the service remains the same by not being instanciated again when route is called.**
 
## Checklist tests
- On postman test @Get("/timeslot/audience/:audienceId") with real audienceId retrieved from the admin console. Then, try with random audience id.
## Issue ticket number and link
[#MVS-114](https://jira.support-ent.fr/browse/MVS-114)
## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/java/fr/openent/viescolaire/security/AccessIfMyStructureCheckIdParam.java [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+import fr.wseduc.webutils.http.Binding;
+import io.vertx.core.Handler;
+import io.vertx.core.http.HttpServerRequest;
+import org.entcore.common.http.filter.ResourcesProvider;
+import org.entcore.common.user.UserInfos;
+
+public class AccessIfMyStructureCheckIdParam implements ResourcesProvider {
+    @Override
+    public void authorize(HttpServerRequest request, Binding binding, UserInfos user, Handler<Boolean> handler) {
+        String structureId = request.params().get("id");
```
Comment (BastienBenard) [inline src/main/java/fr/openent/viescolaire/security/AccessIfMyStructureCheckIdParam.java:12]: Magic String

File: src/main/java/fr/openent/viescolaire/security/AccessIfMyStructureFromAudience.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+import io.vertx.core.http.HttpServerRequest;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import org.entcore.common.http.filter.ResourcesProvider;
+import org.entcore.common.user.UserInfos;
+
+public class AccessIfMyStructureFromAudience implements ResourcesProvider {
+    @Override
+    public void authorize(HttpServerRequest request, Binding binding, UserInfos user, Handler<Boolean> handler) {
+        final Logger log = LoggerFactory.getLogger(Renders.class);
```
Comment (BastienBenard) [inline src/main/java/fr/openent/viescolaire/security/AccessIfMyStructureFromAudience.java:18]: A placer en variable de classe static

File: src/main/java/fr/openent/viescolaire/controller/TimeSlotController.java [hunk 3]
```diff
@@ -140,8 +139,13 @@ public void getDefaultStructureTimeSlot(final HttpServerRequest request) {
         });
     }
 
+    /**
+     * @param request
+     * @queryParam {structureId} mandatory
+     */
     @Put("/time-slots")
```
Comment (BastienBenard) [inline src/main/java/fr/openent/viescolaire/controller/TimeSlotController.java:146]: A voir pour changé en 
```suggestion
    @Put("/structure/:structureId/time-slots")
```
pour respecter le REST API.
Je pense que tu peux faire cette modification puis que de tout façon tu modifie tout les appels a cause du @queryParam {structureId} mandatory

@lapeyronnie qu'est ce que tu en pense ?

File: src/main/java/fr/openent/viescolaire/security/AccessIfMyStructureFromAudience.java (hunk 2 above)
Comment (Gregprod33) [reply src/main/java/fr/openent/viescolaire/security/AccessIfMyStructureFromAudience.java:18]: corrigé

File: src/main/java/fr/openent/viescolaire/controller/TimeSlotController.java [hunk 4]
```diff
@@ -140,8 +139,13 @@ public void getDefaultStructureTimeSlot(final HttpServerRequest request) {
         });
     }
 
+    /**
+     * @param request
+     * @queryParam {structureId} mandatory
+     */
```
[the remark this answers is not in the document]
Comment (Gregprod33) [reply src/main/java/fr/openent/viescolaire/controller/TimeSlotController.java:145]: Pas de doc API pour cette route

File: src/main/java/fr/openent/viescolaire/controller/TimeSlotController.java (hunk 3 above)
Comment (qpic-cgi) [reply src/main/java/fr/openent/viescolaire/controller/TimeSlotController.java:146]: En l'occurrence pour cet endpoint, le body était déjà composé de la structureId avant les modifs.
A voir du coup s'il est judicieux de faire cette modif.

File: src/main/java/fr/openent/viescolaire/security/AccessIfMyStructureCheckIdParam.java [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+import io.vertx.core.http.HttpServerRequest;
+import org.entcore.common.http.filter.ResourcesProvider;
+import org.entcore.common.user.UserInfos;
+
+public class AccessIfMyStructureCheckIdParam implements ResourcesProvider {
+    @Override
+    public void authorize(HttpServerRequest request, Binding binding, UserInfos user, Handler<Boolean> handler) {
+        String structureId = request.params().get(Field.ID);
+        handler.handle(structureId != null && user.getStructures().contains(structureId));
+    }
```
Comment (qpic-cgi) [inline src/main/java/fr/openent/viescolaire/security/AccessIfMyStructureCheckIdParam.java:15]: essai de remplacer tes endpoints par "/structures/:structureId/.../..." (au lieu de "/structures/:id/.../...") comme ça, ça te permet d'utiliser AccessIfMyStructure plutôt, et ça évite de multiplier des checker si ils ont des structure similaires

File: src/main/java/fr/openent/viescolaire/security/AccessIfMyStructureFromAudience.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    public static final Logger log = LoggerFactory.getLogger(Renders.class);
+    @Override
+    public void authorize(HttpServerRequest request, Binding binding, UserInfos user, Handler<Boolean> handler) {
+        String audienceId = request.params().get(Field.AUDIENCEID);
+        ClasseService service = new DefaultClasseService();
+        if (audienceId == null) {
+            handler.handle(false);
+            return;
+        }
+        service.getClasseIdFromAudience(audienceId)
```
Comment (qpic-cgi) [inline src/main/java/fr/openent/viescolaire/security/AccessIfMyStructureFromAudience.java:25]: ça fonctionne ça ? tu peux utiliser ton service dans ton ResourcesProvider  ?

File: src/main/java/fr/openent/viescolaire/security/AccessIfMyStructureFromAudience.java [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+                .onSuccess(etabInfos -> {
+                    if (etabInfos.isEmpty() || etabInfos.getJsonObject(0).isEmpty()) {
+                        handler.handle(false);
+                        return;
+                    }
+                    String structureId = etabInfos.getJsonObject(0).getString(Field.IDSTRUCTURE);
+                    handler.handle(user.getStructures().contains(structureId));
+                })
+                .onFailure(err -> {
+                    log.error("[Viescolaire@TimeSlotController] Failed to retrieve structure from audience", err.getMessage());
```
Comment (qpic-cgi) [inline src/main/java/fr/openent/viescolaire/security/AccessIfMyStructureFromAudience.java:36]: [Viescolaire@AccessIfMyStructureFromAudience::authorize] plutôt

**Review by qpic-cgi — CHANGES_REQUESTED** [review]

File: src/main/java/fr/openent/viescolaire/security/AccessIfMyStructureFromAudience.java (hunk 7 above)
Comment (Gregprod33) [reply src/main/java/fr/openent/viescolaire/security/AccessIfMyStructureFromAudience.java]: Fait

File: src/main/java/fr/openent/viescolaire/security/AccessIfMyStructureCheckIdParam.java (hunk 5 above)
Comment (Gregprod33) [reply src/main/java/fr/openent/viescolaire/security/AccessIfMyStructureCheckIdParam.java]: Oui très bien vu, j'ai modifié

**Review by flmartineau — APPROVED** [review]

File: src/main/java/fr/openent/viescolaire/controller/TimeSlotController.java [hunk 8]
```diff
@@ -165,7 +169,8 @@ public void update(final HttpServerRequest request) {
      */
     @Get("/timeslot/audience/:audienceId")
```
[the remark this answers is not in the document]
Comment (lapeyronnie) [reply src/main/java/fr/openent/viescolaire/controller/TimeSlotController.java:170]: pour aucun souci, competences ne consomme pas cette API

**Review by lapeyronnie — APPROVED** [review]

**Pull request merged**